File: backend/app/schemas/interview.py
"""Schemas for interviews."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.interview import QuestionType, InterviewSessionStatus


class InterviewQuestionBase(BaseModel):
    """Base schema for Interview Question."""
    question_text: str
    question_type: str
    order: int = 0
    target_roles: list[str] | None = None


class InterviewQuestionResponse(InterviewQuestionBase):
    """Schema for Interview Question response."""
    id: UUID
    subcategory_id: UUID
    subcategory_code: str | None = None
    follow_up_on_yes_id: UUID | None = None
    follow_up_on_no_id: UUID | None = None
    is_active: bool

    model_config = {"from_attributes": True}


class InterviewSessionBase(BaseModel):
    """Base schema for Interview Session."""
    interviewee_name: str | None = Field(default=None, max_length=255)
    interviewee_role: str | None = Field(default=None, max_length=255)
    notes: str | None = None


class InterviewSessionCreate(InterviewSessionBase):
    """Schema for creating an Interview Session."""
    assessment_id: UUID
    interviewee_id: UUID | None = None


class InterviewSessionUpdate(BaseModel):
    """Schema for updating an Interview Session."""
    interviewee_name: str | None = Field(default=None, max_length=255)
    interviewee_role: str | None = Field(default=None, max_length=255)
    notes: str | None = None
    status: str | None = None


class InterviewSessionResponse(InterviewSessionBase):
    """Schema for Interview Session response."""
    id: UUID
    assessment_id: UUID
    interviewee_id: UUID | None = None
    status: str
    current_question_index: int
    total_questions: int
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class InterviewResponseBase(BaseModel):
    """Base schema for Interview Response."""
    response_text: str | None = None
    response_value: str | None = Field(default=None, max_length=50)
    confidence_level: str | None = Field(default=None, max_length=50)
    evidence_references: dict | None = None


class InterviewResponseCreate(InterviewResponseBase):
    """Schema for creating an Interview Response."""
    question_id: UUID


class InterviewResponseResponse(InterviewResponseBase):
    """Schema for Interview Response response."""
    id: UUID
    session_id: UUID
    question_id: UUID
    responded_at: datetime
    created_at: datetime
    question_text: str | None = None

    model_config = {"from_attributes": True}


class NextQuestionResponse(BaseModel):
    """Response for next interview question."""
    question: InterviewQuestionResponse | None = None
    question_number: int
    total_questions: int
    progress_percentage: float
    is_complete: bool = False


class InterviewProgressResponse(BaseModel):
    """Response for interview progress."""
    session_id: UUID
    status: str
    questions_answered: int
    total_questions: int
    progress_percentage: float
    responses: list[InterviewResponseResponse] | None = None

File: frontend/lib/types/interview.ts
import { UUID } from './common';

export type InterviewSessionStatus = 'NOT_STARTED' | 'IN_PROGRESS' | 'COMPLETED' | 'PAUSED';

export interface InterviewQuestion {
  id: UUID;
  subcategory_id: UUID;
  subcategory_code: string | null;
  question_text: string;
  question_type: string;
  order: number;
  target_roles: string[] | null;
  follow_up_on_yes_id: UUID | null;
  follow_up_on_no_id: UUID | null;
  is_active: boolean;
}

export interface InterviewSessionCreate {
  assessment_id: UUID;
  interviewee_name?: string;
  interviewee_role?: string;
  notes?: string;
}

export interface InterviewSession {
  id: UUID;
  assessment_id: UUID;
  interviewee_id: UUID | null;
  interviewee_name: string | null;
  interviewee_role: string | null;
  notes: string | null;
  status: string;
  current_question_index: number;
  total_questions: number;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface InterviewResponseCreate {
  question_id: UUID;
  response_text?: string;
  response_value?: string;
  confidence_level?: string;
  evidence_references?: Record<string, unknown>;
}

export interface InterviewResponse {
  id: UUID;
  session_id: UUID;
  question_id: UUID;
  response_text: string | null;
  response_value: string | null;
  confidence_level: string | null;
  evidence_references: Record<string, unknown> | null;
  responded_at: string;
  created_at: string;
  question_text: string | null;
}

export interface NextQuestionResponse {
  question: InterviewQuestion | null;
  question_number: number;
  total_questions: number;
  progress_percentage: number;
  is_complete: boolean;
}

export interface InterviewProgressResponse {
  session_id: UUID;
  status: string;
  questions_answered: number;
  total_questions: number;
  progress_percentage: number;
  responses: InterviewResponse[] | null;
}
